Mathlib/LinearAlgebra/Basis/HasCanonicalBasis.lean

```diff
@@ -0,0 +1,104 @@
+/-
+Copyright (c) 2025 Paul Lezeau. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Paul Lezeau, Eric Wieser
+-/
+import Mathlib.Analysis.InnerProductSpace.PiL2
+
+/-! # Canonical Bases for vector spaces
+
+This file introduces the notion of canonical bases for modules,
+which allows us to work with an implicit notion of the canonical basis for familiar spaces.
+
+-/
+
+universe u v w u' v' w'
+
+/-- `HasCanonicalBasis 𝕜 V ι f` means that the `𝕜`-module `V` is canonically
+represented by an `ι`-indexed basis with elements `f`.
+
+`f` is an outParam so that theorems about a particular basis can use the simp-normal `Pi.single`
+rather than `Pi.basisFun`. -/
+class HasCanonicalBasis (𝕜 : Type u) (V : Type v) (ι : outParam <| Type w)
+    --TODO: can some of these be `semiOutParam`/regular params?
+    (f : outParam <| ι → V) [Semiring 𝕜] [AddCommGroup V]
```

Comment:
I'm really unsure about `f` as an argument here. Why is it not a field of the class?

My general rule of thumb is that if `f` appears in some type in downstream code then it should be a parameter but if it does not appear in the type then it should be a field. 

Having separate `f` field can still be useful for kernel reduction as `basis = f` might be only prepositionally equal but not defeq. However, I'm still a bit skeptical, why doesn't `HasCanonicalBasis` just pick canonical `Basis`? Do you really need good defeq here?